core/src/test/java/com/novoda/noplayer/internal/exoplayer/AdvertStateTest.java

```diff
@@ -0,0 +1,307 @@
+
+import utils.ExceptionMatcher;
+
+import static com.novoda.noplayer.AdvertBreakFixtures.anAdvertBreak;
+import static com.novoda.noplayer.AdvertFixtures.anAdvert;
+import static org.fest.assertions.api.Assertions.assertThat;
+import static org.mockito.BDDMockito.then;
+import static org.mockito.Mockito.mock;
+
+public class AdvertStateTest {
+
+    private static final boolean IS_NOT_PLAYING_ADVERT = false;
+    private static final boolean IS_PLAYING_ADVERT = true;
+    private static final int UNSET_ADVERT_BREAK_INDEX = -1;
+    private static final int UNSET_ADVERT_INDEX = -1;
+    private static final int ONE_SECOND_IN_MILLIS = 1000;
+    private static final int TWO_SECONDS_IN_MILLIS = 2000;
+    private static final int THREE_SECONDS_IN_MILLIS = 3000;
+
+    private static final Advert FIRST_ADVERT = anAdvert()
+            .withDurationInMillis(ONE_SECOND_IN_MILLIS)
+            .build();
+    private static final Advert SECOND_ADVERT = anAdvert()
+            .withDurationInMillis(TWO_SECONDS_IN_MILLIS)
+            .build();
+    private static final Advert THIRD_ADVERT = anAdvert()
+            .withDurationInMillis(THREE_SECONDS_IN_MILLIS)
+            .build();
+
+    private static final AdvertBreak FIRST_ADVERT_BREAK = anAdvertBreak()
+            .withStartTimeInMillis(ONE_SECOND_IN_MILLIS)
+            .withAdvert(FIRST_ADVERT)
+            .build();
+    private static final AdvertBreak SECOND_ADVERT_BREAK = anAdvertBreak()
+            .withStartTimeInMillis(TWO_SECONDS_IN_MILLIS)
+            .withAdverts(FIRST_ADVERT, SECOND_ADVERT)
+            .build();
+    private static final AdvertBreak THIRD_ADVERT_BREAK = anAdvertBreak()
+            .withStartTimeInMillis(THREE_SECONDS_IN_MILLIS)
+            .withAdverts(FIRST_ADVERT, SECOND_ADVERT, THIRD_ADVERT)
+            .build();
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    private final List<AdvertBreak> ADVERT_BREAKS = Arrays.asList(FIRST_ADVERT_BREAK, SECOND_ADVERT_BREAK, THIRD_ADVERT_BREAK);
+    private final AdvertState.Callback callback = mock(AdvertState.Callback.class);
+
+    private final AdvertState advertState = new AdvertState(
+            ADVERT_BREAKS,
+            callback
+    );
+
+    @Test
+    public void handlesPrepareError() {
+        IOException ioException = new IOException("prepare error");
+
+        advertState.handlePrepareError(0, 0, ioException);
+
+        then(callback).should().onHandledPrepareError(0, 0, FIRST_ADVERT, ioException);
+    }
+
+    @Test
+    public void emitsNoEvents_whenNotTransitioningToAdverts() {
+        advertState.update(IS_NOT_PLAYING_ADVERT, UNSET_ADVERT_BREAK_INDEX, UNSET_ADVERT_INDEX);
+
+        then(callback).shouldHaveZeroInteractions();
+    }
+
+    @Test
+    public void notifiesStartOfAdvertBreak_andAdvert_whenTransitioningToAdverts() {
+        advertState.update(IS_PLAYING_ADVERT, 0, 0);
+
+        then(callback).should().onAdvertBreakStart(FIRST_ADVERT_BREAK);
```

Comment:
would it make sense to use `Mockito.inOrder()` for these assertions to ensure that the order doesn't get changed by accident?

Reply:
Huh weird, yea all of these tests should have `inOrder` to them 😬 because the order does actually matter. Thanks for noticing @JozefCeluch I don't know how I managed to not include it 😭